### PR TITLE
Developer

### PR DESCRIPTION
**Pull request by PavelDonchenko**
Create TCP Client
Create TCP Server

### CODE REVIEW

**Comment by penkova** [comment]
"develop" would be a better branch name instead of "developer"

File: tcpS.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer resp.Body.Close()
+
+	xmlCurrenciesData, err := ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		log.Fatal(err)
```
Comment (penkova) [inline tcpS.go:38]: `log.Fatal` will exit, and `defer resp.Body.Close()` will not run.
 Please learn how works `defer` and `log.Fatal(err)` and resolve it.

File: tcpC.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+)
+
+func main() {
+	arguments := os.Args
```
[the remark this answers is not in the document]
Comment (PavelDonchenko) [reply tcpC.go:12]: os.Args provides access to raw command-line arguments, contain an array. Second argument it's:  127.0.0.1:1234 in tcpClient or 1234 in tcpServer.  Or we can write:  l, err := net.Listen("tcp", ":8080") and our server always run on localhost:8080.

File: tcpS.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		clientReq := strings.TrimSpace(string(netData))
+
+		for _, rate := range rates {
+			if rate.Currency == strings.ToUpper(clientReq) {
+				c.Write([]byte(currTime + "  "))
+				c.Write([]byte("1 EURO = " + rate.Rate))
+			}
+
+		}
+		c.Write([]byte(strings.ToUpper(netData)))
```
Comment (PavelDonchenko) [inline tcpS.go:91]: Without line 91 app doesn't work. Why?

File: getCurrency.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+			Currency string `xml:"currency,attr"`
+			Rate     string `xml:"rate,attr"`
+		} `xml:"Cube"`
+	} `xml:"Cube>Cube"`
+}
+
+func GetCurrency() Envelope {
+	// get the latest exchange rate
+	resp, err := http.Get("http://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml")
+
```
Comment (penkova) [inline getCurrency.go:23]: don't use empty line between method and their error check
```
resp, err := http.Get("http://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml")
if err != nil {
		fmt.Println(err)
	}

File: getCurrency.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	defer resp.Body.Close()
+
+	xmlCurrenciesData, err := ioutil.ReadAll(resp.Body)
+
```
Comment (penkova) [inline getCurrency.go:31]: the same

File: getCurrency.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	var env Envelope
+	err = xml.Unmarshal(xmlCurrenciesData, &env)
+
```
Comment (penkova) [inline getCurrency.go:38]: the same

File: tcpS.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+		}
+
+		if strings.TrimSpace(string(netData)) == "STOP" {
+			fmt.Println("Exiting TLC server !")
+			return
+		}
+
+		currTime := GetCurrency().Cube[0].Date
+		rates := GetCurrency().Cube[0].Rates
+		clientReq := strings.TrimSpace(string(netData))
```
Comment (penkova) [inline tcpS.go:47]: Why are you using `string(netData)`? `netData` is already a string :)

Comment (PavelDonchenko) [reply tcpS.go:47]: I didn't know it. That was in example)

**Review by penkova — APPROVED** [review]

**Pull request merged**